File: solution/c#/Day21/Day21/AddRecordUseCase.cs
namespace Day21
{
    public sealed class AddRecordUseCase(string directoryName, AuditManager auditManager, Persister persister)
    {
        public void Handle(string visitorName, DateTime timeOfVisit)
        {
            var files = persister.ReadDirectory(directoryName);

            persister.ApplyUpdate(directoryName,
                auditManager.AddRecord(files, visitorName, timeOfVisit)
            );
        }
    }
}

File: solution/ts/day21/src/addRecordUseCase.ts
import {AuditManager} from './audit';
import {Persister} from "./persister";

export class AddRecordUseCase {
    constructor(private readonly directory: string,
                private readonly auditManager: AuditManager,
                private readonly persister: Persister) {
    }

    public async handle(visitor: string, time: Date): Promise<void> {
        const files = await this.persister.readDirectory(this.directory);
        const update = this.auditManager.addRecord(files, visitor, time);

        await this.persister.applyUpdate(this.directory, update);
    }
}
